File: src/main/java/org/co2dice/mirai/bean/dice/AttributeFixDice.java
package org.co2dice.mirai.bean.dice;

import com.mojang.datafixers.util.Either;
import kotlin.jvm.functions.Function1;
import org.co2dice.mirai.bean.chessman.attribute.*;
import org.co2dice.mirai.bean.chessman.instance.ChessmanInstance;
import org.co2dice.mirai.bean.counter.Counter;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.stream.Collectors;

/**
 * @author DUELIST
 * 这个类的目的是，将抽象的属性修正值转换为具体的骰子修正值，比如将玩家的体质转化为体质数值的骰子
 * 可能存在多个属性的修正值，比如玩家的体质/2+意志/2
 *
 */
public class AttributeFixDice {

    private final Function1<AttributeInstanceTable,Either<DiceList,String>> fixFunc;
    //获取属性修正值的函数

    public AttributeFixDice( Function1<AttributeInstanceTable,Either<DiceList,String>> fixFunc) {
        this.fixFunc = fixFunc;
    }


    public AttributeFixDice(AttributeAPI a) {
        this.fixFunc = (table -> {
            Integer value = table.getValue(a);
            if( value != null ) {
                return Either.left(new DiceList(value));
            }
            return Either.right("属性修正值为空,为空属性:" + a.getNameStr());
        });
        //默认的修正值,传入属性，返回那个属性的数值
        //如果属性不存在会报错，需要在调用的最外圈加一层try catch
    }

    public boolean canUse(AttributeInstanceTable table){
        return fixFunc.invoke(table).left().isPresent();
    }

    @NotNull
    public Either<DiceList,String> getDiceList(ChessmanInstance c){
        //通过玩家获取属性修正值
        var table = c.getAttributeInstanceTable();
        return fixFunc.invoke(table);
    }

    public Either<List<Dice>,String> getListDice(ChessmanInstance c){
        //通过玩家获取属性修正值
        var either = getDiceList(c);
        if(either.left().isPresent()){
            return Either.left(either.left().get().getDiceList());
        }else if (either.right().isPresent()){
            return Either.right(either.right().get());
        }
        return Either.right("未知错误");
    }


    public Set<AttributeAPI> getNeeds() {
        //获取这个修正值需要的属性,使用遍历法
        Set<AttributeAPI> set = new java.util.HashSet<>(Set.of());
        set.addAll(Arrays.stream(EliteAttribute.values()).toList());
        set.addAll(Arrays.stream(MobAttribute.values()).toList());
        //现在这个set里获取了所有类型的属性，无论是不是怪物的

        set.removeIf(a -> {
            //将set里的属性全部转化为AttributeInstance，然后填充入table中
            var table = new AttributeInstanceTable(
                    set.stream().filter(i -> !i.equals(a))
                            .map(i -> new AttributeInstance(i,2,2)).collect(Collectors.toSet())
            );
            return canUse(table);
            //若属性依然可以检定，那么就是需要的属性
        });
        return set;
    }

    public Either<Integer,String> roll(ChessmanInstance c){
        var either = getDiceList(c);
        if(either.left().isPresent()){
            return Either.left(either.left().get().roll().getResult());
        }else if (either.right().isPresent()){
            return Either.right(either.right().get());
        }
        return Either.right("检查属性时发生未知错误");
    }
}

File: src/main/java/org/co2dice/mirai/bean/dice/AttributeFixDice.kt
package org.co2dice.mirai.bean.dice

import com.mojang.datafixers.util.Either
import org.co2dice.mirai.bean.chessman.attribute.*
import org.co2dice.mirai.bean.chessman.instance.ChessmanInstance
import java.util.*
import java.util.stream.Collectors

/**
 * @author DUELIST
 * 这个类的目的是，将抽象的属性修正值转换为具体的骰子修正值，比如将玩家的体质转化为体质数值的骰子
 * 可能存在多个属性的修正值，比如玩家的体质/2+意志/2
 */
class AttributeFixDice {
    private val fixFunc: Function1<AttributeInstanceTable, Either<DiceList, String>>

    //获取属性修正值的函数
    constructor(fixFunc: Function1<AttributeInstanceTable?, Either<DiceList, String>>) {
        this.fixFunc = fixFunc
    }

    constructor(a: AttributeAPI) {
        fixFunc = label@{ table: AttributeInstanceTable ->
            val value = table.getValue(a)
            if (value != null) {
                return@label Either.left<DiceList, String>(DiceList(value))
            }
            Either.right("属性修正值为空,为空属性:" + a.nameStr)
        }
        //默认的修正值,传入属性，返回那个属性的数值
        //如果属性不存在返回right
    }

    fun canUse(table: AttributeInstanceTable): Boolean {
        return fixFunc(table).left().isPresent
    }

    fun getDiceList(c: ChessmanInstance): Either<DiceList, String> {
        //通过玩家获取属性修正值
        val table = c.attributeInstanceTable
        return fixFunc.invoke(table)
    }

    fun getListDice(c: ChessmanInstance): Either<List<Dice>, String> {
        //通过玩家获取属性修正值
        val either = getDiceList(c)
        if (either.left().isPresent) {
            return Either.left(either.left().get().diceList)
        } else if (either.right().isPresent) {
            return Either.right(either.right().get())
        }
        return Either.right("未知错误")
    }//将set里的属性全部转化为AttributeInstance，然后填充入table中

    //获取这个修正值需要的属性,使用遍历法
    fun getNeeds(): Set<AttributeAPI> {
            //获取这个修正值需要的属性,使用遍历法
            val set: MutableSet<AttributeAPI> = HashSet()
            set.addAll(Arrays.stream(EliteAttribute.values()).toList())
            set.addAll(Arrays.stream(MobAttribute.values()).toList())
            //现在这个set里获取了所有类型的属性，无论是不是怪物的
            set.removeIf { a: AttributeAPI ->
                //将set里的属性全部转化为AttributeInstance，然后填充入table中
                val table = AttributeInstanceTable(
                    set.stream().filter { i: AttributeAPI -> i != a }
                        .map { i: AttributeAPI? ->
                            AttributeInstance(
                                i!!, 8, 8
                            )
                        }.collect(Collectors.toSet())
                )
                canUse(table)
            }
            return set
        }

    fun roll(c: ChessmanInstance): Either<Int, String> {
        val either = getDiceList(c)
        if (either.left().isPresent) {
            return Either.left(either.left().get().roll().getResult())
        } else if (either.right().isPresent) {
            return Either.right(either.right().get())
        }
        return Either.right("检查属性时发生未知错误")
    }
}